File: 149.max-points-on-a-line.java
/*
 * @lc app=leetcode id=149 lang=java
 *
 * [149] Max Points on a Line
 */
class Solution {

    public int maxPoints(int[][] points) {
        int res = 0;
        if (points == null || points.length == 0) return res;
        for (int i = 0; i < points.length; ++i) {
            Map<Slope, Integer> slopeMap = new HashMap<>();
            int coincidence = 0;
            for (int j = i + 1; j < points.length; ++j) {
                Slope key = new Slope(points[i], points[j]);
                if (Slope.COINCIDENCE_KEY.equals(key)) coincidence++;
                else slopeMap.put(key, slopeMap.getOrDefault(key, 0) + 1);
            }
            if (coincidence > res) res = coincidence;
            for (int count : slopeMap.values())
                if (count+coincidence > res) res = count+coincidence;
        }
        return res + 1;
    }

    static class Slope {

        public static final Slope COINCIDENCE_KEY = new Slope(0, 0);

        final int dx, dy;
        public Slope(int[] p1, int[] p2) {
            int tx = p1[0]-p2[0];
            int ty = p1[1]-p2[1];
            int gcdxy = gcd(tx, ty);
            if (gcdxy != 0) {
                tx /= gcdxy;
                ty /= gcdxy;
                dx = tx < 0 ? -tx : tx;
                dy = tx < 0 ? -ty : ty;
            } else {
                dx = 0;
                dy = 0;
            }
        }

        public Slope(int dx, int dy) {
            this.dx = dx;
            this.dy = dy;
        }

        private int gcd(int a, int b) {
            if (b == 0) return a;
            return gcd(b, a % b);
        }

        @Override
        public boolean equals(Object oo) {
            Slope o = (Slope) oo;
            return dx == o.dx && dy == o.dy;
        }

        @Override
        public int hashCode() {
            return dx * dy * 7 + 13;
        }
    }
}

File: go/149.max-points-on-a-line.go
package main

/*
 * @lc app=leetcode id=149 lang=golang
 *
 * [149] Max Points on a Line
 */

// @lc code=start

type slope struct {
	xl, yl int // relatively prime
}

type line struct {
	slope
}

var noSlope = slope{0, 0}

func maxPoints(points [][]int) int {
	res := 0
	if len(points) == 0 {
		return res
	}

	for i := 0; i < len(points); i++ {
		slopes := make(map[slope]int)
		for j := i + 1; j < len(points); j++ {
			slopes[newSlope(points[i], points[j])]++
		}
		coincide := slopes[noSlope]
		if res < coincide {
			res = coincide
		}
		slopes[noSlope] = 0
		for _, c := range slopes {
			if res < c+coincide {
				res = c + coincide
			}
		}
	}
	return res + 1
}

func newSlope(pi, pj []int) slope {
	yl := pj[1] - pi[1]
	xl := pj[0] - pi[0]
	d := gcd(yl, xl)
	if d != 0 {
		xl, yl = xl/d, yl/d
		if xl < 0 {
			xl, yl = -xl, -yl
		}
		return slope{xl, yl}
	}
	return slope{0, 0}
}

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}

// @lc code=end
